### PR TITLE
Model selection

### PR DESCRIPTION
**Pull request by niedakh**
Introducing the model selection subpackage and iterative stratification algorithms

### CODE REVIEW

File: skmultilearn/model_selection/iterative_stratification.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            row = self.available_samples.pop()
+            self.rows_used[row] = True
+            self.samples_left -= 1
+            fold_selected = np.random.choice(
+                np.where(self.desired_samples_per_fold > 0)[0],
+                1)[0]
+            self.desired_samples_per_fold[fold_selected] -= 1
+            self.folds[fold_selected].append(row)
+
+    def _iter_test_indices(self, X, y=None, groups=None):
```
Comment (ljvmiranda921) [inline skmultilearn/model_selection/iterative_stratification.py]: Hmmm, please correct me if my understanding is wrong, but I assume that `_iter_test_indices` is like the
main routine that calls all other subroutines such as `distribute_positive_evidence` and `distribute_negative_evidence`.
I believe that the latter two methods will not be called by the user. They're just there to compartmentalize important parts of the algorithm.

My question is, is `_iter_test_indices` a user-facing method? If it's being called from the API, then we should remove the underscore before it. In my opinion, I think the one that should have underscores prepended their names would be all of the `init_*`  methods, the `distrubute_*` methods, and the label tie-in. 

In addition, if `_iter_test_indices` is a user-facing method, then it is necessary to put some docstrings to explain how their parameters.

File: skmultilearn/model_selection/iterative_stratification.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                        self.samples_with_combination[i].remove(row)
+                    self.desired_samples_per_label_per_fold[i][m] -= 1
+                self.desired_samples_per_fold[m] -= 1
+
+            l = self.get_most_desired_combination()
+
+    def distribute_negative_evidence(self):
+        self.available_samples = [
+            i for i, v in self.rows_used.items() if not v]
+        self.samples_left = len(self.available_samples)
```
Comment (ljvmiranda921) [inline skmultilearn/model_selection/iterative_stratification.py]: Maybe @ChristianSch  can also weigh-in on this. But I find that assigning `self.samples_left` as an attribute is a bit costly. I see the reasoning for it, because we're deducting `-1` per iteration. But maybe we can write it this way:

```python
for i in range(len(self.available_samples)):
    #do something
```

In this case, we're relieve with the burden of decrementing `self.samples_left`, and we don't need to assign an attribute into it.

File: skmultilearn/model_selection/iterative_stratification.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        if len(M) == 1:
+            return M[0]
+        else:
+            max_val = max(self.desired_samples_per_fold[M])
+            M_prim = np.where(
+                np.array(self.desired_samples_per_fold) == max_val)[0]
+            M_prim = np.array([x for x in M_prim if x in M])
+            return np.random.choice(M_prim, 1)[0]
+
+    def get_most_desired_combination(self):
```
Comment (ljvmiranda921) [inline skmultilearn/model_selection/iterative_stratification.py]: I think this can be written in such a way that `currently_chosen` is always returned instead of having two return statements. Since the initial value of `currently_chosen` is `None`, then we just return this initial value given these conditions and return another value given another set of conditions. I think this may be a better practice than returning the actual `None` itself.

This means that everything inside the `for-loop` corresponds to changing the value of whatever `currently_chosen` is.

File: skmultilearn/model_selection/iterative_stratification.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+            fold_selected = np.random.choice(
+                np.where(self.desired_samples_per_fold > 0)[0],
+                1)[0]
+            self.desired_samples_per_fold[fold_selected] -= 1
+            self.folds[fold_selected].append(row)
+
+    def _iter_test_indices(self, X, y=None, groups=None):
+        self.init_values(X)
+        if self.random_state:
+            check_random_state(self.random_state)
```
[the remark this answers is not in the document]
Comment (ChristianSch) [reply skmultilearn/model_selection/iterative_stratification.py]:
```
from sklearn.utils import check_random_state
```

this should be what's missing.

File: skmultilearn/model_selection/iterative_stratification.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+    :param n_splits: with number of splits
+    :type n_splits: int
+
+    :param order: taking into account order-th order of relationship
+    :type order: int
+
+    :param random_state: and the random state seed (optional)
+    :type random_state: int
+
```
[the remark this answers is not in the document]
Comment (ChristianSch) [reply skmultilearn/model_selection/iterative_stratification.py]: I agree, it's important to have a consistent style when it comes to documentation. We should clarify what we want to use in the future and refactor the whole project. (Follow-up issue maybe?)

File: skmultilearn/model_selection/iterative_stratification.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (niedakh) [reply skmultilearn/model_selection/iterative_stratification.py]: fixed

File: skmultilearn/model_selection/iterative_stratification.py (hunk 1 above)
Comment (niedakh) [reply skmultilearn/model_selection/iterative_stratification.py]: I've made distribute* methods private (start with _), also _iter_test_indices is not a user-facing method, the _iter_test_indices is used internally by the scikit BaseKFold class, the only user-facing method we have here is the init.

File: skmultilearn/model_selection/iterative_stratification.py (hunk 2 above)
Comment (niedakh) [reply skmultilearn/model_selection/iterative_stratification.py]: I agree, fixed.

File: skmultilearn/model_selection/iterative_stratification.py (hunk 3 above)
Comment (niedakh) [reply skmultilearn/model_selection/iterative_stratification.py]: done.

File: skmultilearn/model_selection/iterative_stratification.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    to handle high-order relationships in the data set, if order = 1, the algorithm falls back to the original Sechidis11 setting. 
+
+    If order is larger than 1 this class constructs a list of label combinations with replacement, i.e. allowing combinations of lower
+    order to be take into account. For example for combinations of order 2, the stratifier will consider both
+    label pairs (1, 2) and single labels denoted as (1,1) in the algorithm. In higher order cases the 
+    when two combinations of different size have similar desirablity: the larger, i.e. more specific combination
+    is taken into consideration first, thus if a label pair (1,2) and label 1 represented as (1,1) are of similar
+    desirability, evidence for (1,2) will be assigned to folds first.
+
+
```
[the remark this answers is not in the document]
Comment (niedakh) [reply skmultilearn/model_selection/iterative_stratification.py]: added the example!

**Comment by niedakh** [comment]
@ChristianSch I've documented internal and helper functions.
@ljvmiranda921 I've adhered to your review, there's less bloat in class properties, private functions are made private, helpers are moved out of the class, lot's of docs added. 

Please take a look at it, and approve the PR if it's ok, or comment if not :)
